### PR TITLE
Bug/csi 3586 expand nvme bad path

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

File: node/pkg/driver/node_utils.go [hunk 1]
```diff
@@ -177,6 +179,23 @@ func (n NodeUtils) StageInfoFileIsExist(filePath string) bool {
 	}
 	return true
 }
+
+func (n NodeUtils) DevicesAreNvme(sysDevices []string) bool {
+	args := []string{"list"}
+	out, err := n.Executer.ExecuteWithTimeout(TimeOutMultipathdCmd, nvmeCmd, args)
```
Comment (oriyarde) [inline node/pkg/driver/node_utils.go:185]: could we use a different timeout const here?
if you prefer, we can also have more general timeout consts, to use in specific timeout consts definition

File: node/pkg/driver/node_utils.go [hunk 2]
```diff
@@ -177,6 +179,23 @@ func (n NodeUtils) StageInfoFileIsExist(filePath string) bool {
...
+	if err != nil {
+		return false
```
Comment (oriyarde) [inline node/pkg/driver/node_utils.go:187]: what if the command fails, but they are actually nvme devices?

File: node/pkg/driver/node_utils.go [hunk 3]
```diff
@@ -177,6 +179,23 @@ func (n NodeUtils) StageInfoFileIsExist(filePath string) bool {
...
+	}
+	devices := string(out)
+	for _, deviceName := range sysDevices {
+		if strings.Contains(devices, deviceName) {
+			logger.Debugf("found device {%s} in nvme list", deviceName)
```
Comment (oriyarde) [inline node/pkg/driver/node_utils.go:192]:
```suggestion
	nvmeDevices := string(out)
	for _, deviceName := range sysDevices {
		if strings.Contains(nvmeDevices, deviceName) {
			logger.Debugf("found device {%s} in nvme list", deviceName)
```

File: node/pkg/driver/node_utils.go [hunk 4]
```diff
@@ -177,6 +179,23 @@ func (n NodeUtils) StageInfoFileIsExist(filePath string) bool {
...
+
+func (n NodeUtils) DevicesAreNvme(sysDevices []string) bool {
+	args := []string{"list"}
+	out, err := n.Executer.ExecuteWithTimeout(TimeOutMultipathdCmd, nvmeCmd, args)
+	if err != nil {
+		return false
+	}
+	devices := string(out)
+	for _, deviceName := range sysDevices {
+		if strings.Contains(devices, deviceName) {
```
Comment (oriyarde) [inline node/pkg/driver/node_utils.go:191]: which of the "lists" is potentially longer?
I think we should check inside the one that is expected to be shorter

File: node/pkg/driver/node_utils.go (hunk 1 above)
Comment (ArbelNathan) [reply node/pkg/driver/node_utils.go:185]: sure,
how do you suggest we call it? `TimeOutGeneralCmd`?

Comment (oriyarde) [reply node/pkg/driver/node_utils.go:185]: if we only have mainly 1 "default" value, then sure

File: node/pkg/driver/node_utils.go (hunk 4 above)
Comment (ArbelNathan) [reply node/pkg/driver/node_utils.go]: sysDevices is way shorter

File: node/pkg/driver/node_utils.go (hunk 2 above)
Comment (ArbelNathan) [reply node/pkg/driver/node_utils.go:187]: that the thing, the user dont have to have nvme packages on its env.. any suggestions?

File: node/pkg/driver/node_utils.go (hunk 1 above)
Comment (ArbelNathan) [reply node/pkg/driver/node_utils.go]: done

File: node/pkg/driver/node_utils.go (hunk 2 above)
Comment (oriyarde) [reply node/pkg/driver/node_utils.go:187]: maybe check if the command exists before?

File: node/pkg/driver/node_utils.go (hunk 1 above)
Comment (oriyarde) [reply node/pkg/driver/node_utils.go]: I didn't mean to replace existing specific timeout consts.
i meant we could have a more general const in addition.
e.g. (assuming you meant to have this value for both commands):
```
TimeOutGeneralCmd           = 10 * 1000
TimeOutMultipathdCmd        = TimeOutGeneralCmd
TimeOutNvmeCmd              = TimeOutGeneralCmd
```

File: node/pkg/driver/node_utils.go (hunk 4 above)
Comment (oriyarde) [reply node/pkg/driver/node_utils.go]: then I think it is more suitable for `Contains`

Comment (oriyarde) [reply node/pkg/driver/node_utils.go]: then again, the first sys device should be contained in nvme devices, so I'm not sure why we want to potentially go over all the volume sys devices

Comment (oriyarde) [reply node/pkg/driver/node_utils.go]: should we skip the rescan if **any** of them is nvme, or **all**?

Comment (ArbelNathan) [reply node/pkg/driver/node_utils.go]: it should find it on the first check, (if it checks nvmeDevices Contains sysDevice). no reason for going over all  the sysDevices

Comment (oriyarde) [reply node/pkg/driver/node_utils.go]: so do we need a for loop?
and if not, should the function even require a **list** of sys devices?

Comment (oriyarde) [reply node/pkg/driver/node_utils.go]: maybe if we have some sort of hybrid-connectivity configuration per volume, we should go over all sys devices before deciding to entirely skip the rescan?

Comment (oriyarde) [reply node/pkg/driver/node_utils.go]: I'm not sure which of the scenarios is even possible, hope you're right, resolving

File: node/pkg/driver/node_utils.go (hunk 1 above)
Comment (ArbelNathan) [reply node/pkg/driver/node_utils.go]: done

File: node/pkg/driver/node_utils.go (hunk 2 above)
Comment (ArbelNathan) [reply node/pkg/driver/node_utils.go]: done

File: node/pkg/driver/node_utils.go [hunk 5]
```diff
@@ -177,6 +181,26 @@ func (n NodeUtils) StageInfoFileIsExist(filePath string) bool {
 	}
 	return true
 }
+
+func (n NodeUtils) DevicesAreNvme(sysDevices []string) (bool, error) {
+	args := []string{"list"}
+	if err := n.Executer.IsExecutable(nvmeCmd); err != nil {
+		return false, nil
```
Comment (oriyarde) [inline node/pkg/driver/node_utils.go:188]:
```suggestion
		return false, err
```

**Review by oriyarde — APPROVED** [review]
LGTM

**Pull request merged**